libraries/type-length-value/src/state.rs

```diff
@@ -2,16 +2,25 @@
 
 use {
     crate::{
-        discriminator::{Discriminator, TlvDiscriminator},
         error::TlvError,
         length::Length,
         pod::{pod_from_bytes, pod_from_bytes_mut},
     },
     bytemuck::Pod,
     solana_program::program_error::ProgramError,
+    spl_discriminator::{Discriminator, SplDiscriminator},
```

Comment:
Seeing these two imports together confuses me a bit. Like, what's the difference between `Discriminator` and `SplDiscriminator`? I have to read back and remember which is the type and which is the trait. Maybe we can rename `SplDiscriminator` to something like `SplTypeDiscriminator` since it's meant to discriminate between concrete types. What do you think?

Reply:
I guess this seems off-putting because of the way we're using the traits combined with `TlvDiscriminator`. Ideally, if you just want a discriminator and don't care about the traits, you can just use the `Discriminator` struct on its own. Which directly relates to your other comment of providing a `new_with_hash_input` function.

The tricky part there is it's not going to be a constant function, so the macro will still have to do the hashing as it currently is, however that still doesn't require a `solana-program` import

Reply:
Thoughts on `HasSplDiscriminator`? or just `HasDiscriminator`?

Reply:
I looked up for any naming conventions for Rust traits, and nothing really came up except for https://github.com/rust-lang/api-guidelines/discussions/28 which would lean toward `Discriminate` as the imperative form, which is pretty hilarious.

And yet, I can't come up with anything better than `SplDiscriminate`. I'm not a fan of `Has*` or `With*` since it's not really canonical anywhere else. Or we can go with `SplDiscriminates`, similar to https://doc.rust-lang.org/rust-by-example/generics/assoc_items/types.html

On another side, maybe we can rename `Discriminator` to `ArrayDiscriminator` to make it clearer that it's a concrete type?

Reply:
I like `ArrayDiscriminator` yep.
I wanted to make a joke about `*Discriminates` but the more I say/look at it the more it kinda makes sense, but it still does have some obvious connotations

Reply:
I think I'm going with `SplDiscriminates`